pkg/codegen/schema/schema.go

```diff
@@ -1129,33 +1181,75 @@ func (t *types) bindObjectType(token string, spec ObjectTypeSpec) (*ObjectType,
 }
 
-func bindTypes(pkg *Package, objects map[string]ObjectTypeSpec) (*types, error) {
+func (t *types) bindEnumTypeDetails(enum *EnumType, token string, spec ComplexTypeSpec) error {
+	typ, err := t.bindType(TypeSpec{Type: spec.Type})
+	if err != nil {
+		return err
+	}
+	enum.Token = token
+	enum.Elements = t.bindEnumValues(spec.Enum)
+	enum.ElementType = typ
+	enum.Comment = spec.Description
+
+	return nil
+}
+
+func (t *types) bindEnumValues(values []*EnumValueSpec) (enums []*Enum) {
+	for _, spec := range values {
+		enum := &Enum{
+			// TODO: check that the type of the value matches the enum type.
+			// https://github.com/pulumi/pulumi/issues/5387
+			Value:   spec.Value,
+			Comment: spec.Description,
+			Name:    spec.Name,
+		}
+		enums = append(enums, enum)
+	}
+	return
+}
+
+func (t *types) bindEnumType(token string, spec ComplexTypeSpec) (*EnumType, error) {
+	enum := &EnumType{}
+	if err := t.bindEnumTypeDetails(enum, token, spec); err != nil {
+		return nil, err
+	}
+	return enum, nil
+}
+
+func bindTypes(pkg *Package, complexTypes map[string]ComplexTypeSpec) (*types, error) {
 	typs := &types{
 		pkg:     pkg,
 		objects: map[string]*ObjectType{},
 		arrays:  map[Type]*ArrayType{},
 		maps:    map[Type]*MapType{},
 		unions:  map[string]*UnionType{},
 		tokens:  map[string]*TokenType{},
-	}
-
-	// Declare object types before processing properties.
-	for token, spec := range objects {
-		if spec.Type != "object" {
-			return nil, errors.Errorf("type %s must be an object, not a %s", token, spec.Type)
+		enums:   map[string]*EnumType{},
+	}
+
+	// Declare object and enum types before processing properties.
+	for token, spec := range complexTypes {
+		if spec.Type == "object" {
+			// It's important that we set the token here. This package interns types so that they can be equality-compared
+			// for identity. Types are interned based on their string representation, and the string representation of an
+			// object type is its token. While this doesn't affect object types directly, it breaks the interning of types
+			// that reference object types (e.g. arrays, maps, unions)
+			typs.objects[token] = &ObjectType{Token: token}
+		} else if len(spec.Enum) > 0 {
+			typs.enums[token] = &EnumType{Token: token}
 		}
-
-		// It's important that we set the token here. This package interns types so that they can be equality-compared
-		// for identity. Types are interned based on their string representation, and the string representation of an
-		// object type is its token. While this doesn't affect object types directly, it breaks the interning of types
-		// that reference object types (e.g. arrays, maps, unions)
-		typs.objects[token] = &ObjectType{Token: token}
 	}
 
 	// Process properties.
-	for token, spec := range objects {
-		if err := typs.bindObjectTypeDetails(typs.objects[token], token, spec); err != nil {
-			return nil, errors.Wrapf(err, "failed to bind type %s", token)
+	for token, spec := range complexTypes {
```

Comment:
They are pre-existing but my impression from the comments is that we want to declare all the object tokens first before processing their properties since the properties might otherwise reference object types that haven't been declared yet? I guess this isn't really relevant for enums so those could be processed fully in the first loop. @pgavlin am I understanding this correctly?

Reply:
That's correct, yes.

Reply:
(future changes will move this to a depth-first approach to accommodate #5385)